.github/workflows/canary.yml

```diff
@@ -0,0 +1,48 @@
+name: canary
+on:
+  - pull_request
+  - push
+jobs:
+  canary:
+    name: testing ${{matrix.package}} with ${{matrix.node}} on ${{matrix.os}}
+    runs-on: ${{matrix.os}}
+    steps:
+      - name: checkout micromark
+        uses: actions/checkout@v2
+      - name: setup node
+        uses: dcodeIO/setup-node-nvm@master
+        with:
+          node-version: ${{matrix.node}}
+      - name: update npm
+        run: npm install -g npm
+      - name: install micromark
+        run: npm install
+      - name: build micromark
+        run: npm run build
+
+      - name: checkout ${{matrix.package}}
+        uses: actions/checkout@v2
+        with:
+          repository: ${{matrix.package}}
+          path: canary/${{matrix.package}}
+      - name: test ${{matrix.package}}
+        run: |
+          npm install
+          npx rimraf "**/micromark"
+          npx rimraf "**/micromark-util-*"
```

Comment:
What is being removed here? Are there `micromark` files or folders in `micromark/canary/remarkjs/remark/`?
Why just `micromark` itself and `micromark-util-*`, but not other `micromark-*` files/folders?

Reply:
copying from earlier in the discussion https://github.com/micromark/micromark/pull/106#discussion_r767339359

> this removes the pinned versions of these libraries that npm installed, meaning node will continue searching upward, until it finds the parent `micromark/node_modules` folder with the just-built version of micromark.

---

> Are there micromark files or folders in micromark/canary/remarkjs/remark/?

There are, this runs right after `npm install` in the _micromark/canary/remarkjs/remark_ folder, so it has a frozen/published copy of micromark installed, we want it to find the latest, just built, copy instead.

> Why just micromark itself and micromark-util-*, but not other micromark-* files/folders?

It seemed to work, open to suggestions for a better remove pattern to get all the micromark packages in this repo cleared at once.

Reply:
ahhh, forgot that I already asked it.
Wouldn’t `npx rimraf "node_modules/micromark*"` do the same but catch everything in [`packages/`](https://github.com/micromark/micromark/tree/main/packages)? (or `npx rimraf "node_modules/{micromark,micromark*}"` if the asterisk matches one or more, which I don’t think it does?)

Reply:
One possible alternative, something like

```
rm -rf `npm ls --workspaces --json | jq "[.dependencies | keys[]] | \"**/\" + .[]"`
```

which tries to get at the parent workspace, list the packages installed there, create a `**/{package name}` pattern for each, then passing all that back to `rm -rf`

Reply:
Hmm, but what about packages needed in the canary packages though? `micromark-extension-mdx`?
Problem is that this repo houses `micromark-factory-*` too, but those extensions can also use `micromark-*` packages? 🤔

Reply:
> Wouldn’t `npx rimraf "node_modules/micromark*"` do the same but catch everything in packages/?

it would! but, it may be too aggressive, also removing things like `micromark-extension-gfm` which may be needed by unit tests and not part of the parent micromark repo.

Reply:
Would `rm -rf "canary/.../node_modules/${ls packages/}"` do the same, but cleaner?

Reply:
which admittedly can even happen with my current pattern.
For example

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'micromark-util-events-to-acorn' imported from /home/runner/work/micromark/micromark/canary/micromark/micromark-extension-mdx/node_modules/micromark-factory-mdx-expression/dev/index.js
    at new NodeError (node:internal/errors:371:5)
    at packageResolve (node:internal/modules/esm/resolve:864:9)
    at moduleResolve (node:internal/modules/esm/resolve:910:18)
    at defaultResolve (node:internal/modules/esm/resolve:1005:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
    at link (node:internal/modules/esm/module_job:78:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

https://github.com/micromark/micromark/runs/4627735817?check_suite_focus=true#step:6:36

Reply:
> Would `rm -rf "canary/.../node_modules/${ls packages/}"` do the same, but cleaner?

potentially, I think we want the reverse of this though. :thinking: 
inside the canary folder, we want to list the packages in the parent micromark folder, and use that list to remove frozen/published copies in the canary folder.

Reply:
Huh, that’s what my suggestion does, right?
- List all packages maintained in the micromark monorepo (which are hoisted to the monorepo `node_modules/`?)
- Install an example repo normally, separately from the monorepo
- Remove what monorepo packages from the example repo `node_modules`

Reply:
Though my example was assuming a CWD of the monorepo, from an example repo my example would be, say, `rm -rf "./node_modules/${ls ../../../packages/}"`

Reply:
testing locally I get `bad substitution`.
But parsing out your intent.

This script is running inside the canary/{package-name} folder.
so `ls packages` would list package inside the `canary` folder right?
`canary` would then try to go into a folder that doesn't exist.
`...` would take it up two levels to the parent micromark folder
`node_modules` would go into the parent's `node_modules` directory

Reply:
resolved in  42e5976